src/server/data_source.rs

```diff
@@ -1,142 +1,197 @@
-use std::ffi::c_void;
+use std::{
+    ffi::c_void,
+    panic::{catch_unwind, AssertUnwindSafe},
+    ptr::NonNull,
+};
 
 use open62541_sys::{
     UA_Boolean, UA_DataSource, UA_DataValue, UA_NodeId, UA_NumericRange, UA_Server, UA_StatusCode,
 };
 
 use crate::{server::NodeContext, ua, DataType};
 
-/// Read callback for [`DataSource`].
-#[allow(clippy::module_name_repetitions)]
-pub type DataSourceRead = dyn FnMut(&mut DataSourceReadContext) -> DataSourceResult;
-
-/// Write callback for [`DataSource`].
 #[allow(clippy::module_name_repetitions)]
-pub type DataSourceWrite = dyn FnMut(&mut DataSourceWriteContext) -> DataSourceResult;
+pub type DataSourceResult = std::result::Result<(), ua::StatusCode>;
 
 /// Data source with callbacks.
 ///
-/// The read and write callbacks implement the operations on the data source when it is added as a
-/// variable node via [`Server::add_data_source_variable_node()`].
+/// The read and write callbacks implement the operations on the variable when it is added via
+/// [`Server::add_data_source_variable_node()`].
 ///
 /// [`Server::add_data_source_variable_node()`]: crate::Server::add_data_source_variable_node
-pub struct DataSource {
-    read: Box<DataSourceRead>,
-    write: Option<Box<DataSourceWrite>>,
+pub trait DataSource {
```

Comment:
Not sure if it would be possible to split this into separate `DataSource`/`DataSourceRead`/`DataSourceWrite` traits?

A single `DataSource` trait is easier to handle. Especially when dealing with _trait objects_.